### PR TITLE
einsum functor

### PR DESCRIPTION
**Pull request by wangyuyue**
This enables users to construct complex tensor computations given the text description in Einstein summation format.
TODO: some steps in the current implementation need new operators' support, e.g. diagonal operation for any two dimensions.

### CODE REVIEW

File: oneflow/core/functional/impl/einsum.cpp [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+    std::shared_ptr<Tensor> result;
+
+    // Fast path for when an operand has zero sized dim
+    if (has_zero_size_dim) {
+      std::vector<int64_t> out_shape(out_size);
+      for (auto i = 0; i < out_size; i++) {
+        out_shape[i] = permuted_operands[dim_last_op[i]]->dim(i);
+      }
+      return std::shared_ptr<Tensor>();
+      // TODO: need function supporting construct zeros tensor with shape
```
Comment (jackalcooper) [inline oneflow/core/functional/impl/einsum.cpp:326]: 用 `ConstantFunctor` 可以吗

File: oneflow/core/functional/functional_api.yaml [hunk 2]
```diff
@@ -1049,6 +1049,10 @@
                                      SbpList sbp_tuple, DataType dtype=None, Generator generator=None)"
   bind_python: True
 
+- name: "einsum"
+  signature: "Tensor EinSum(String einsum_formula, TensorTuple inputs)"
```
Comment (hjchen2) [inline oneflow/core/functional/functional_api.yaml:1053]: 注意一下新版functional接口的规范，https://github.com/Oneflow-Inc/oneflow/wiki/Functional-Interface，改成下面的定义，
```yaml
  signature: "Tensor (String einsum_formula, TensorTuple inputs) => EinSum"
```

File: oneflow/core/functional/impl/einsum.cpp [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+            --dim;
+          } else {
+            sum_dims.push_back(dim);
+          }
+        }
+      }
+
+      // Multiply tensors and sum out dimensions in sum_dims
+      if (sum_dims.empty()) {
+        result = JUST(Multiply(result, operand));
```
Comment (hjchen2) [inline oneflow/core/functional/impl/einsum.cpp:364]: Multiply被重命名为Mul了

**Comment by SunDoge** [comment]
Is there any progress?

**Comment by yuanms2** [comment]
我们这个einsum 的完成度怎么样，还差多少工作量，感觉爱因斯坦求和用处还挺多的，alphafold 的实现就用了

**Comment by Ldpe2G** [comment]
@SunDoge oneflow already support einsum, you can have a try. https://github.com/Oneflow-Inc/oneflow/pull/7526

**Comment by SunDoge** [comment]
@Ldpe2G hi, I have already noticed your excellent pr yesterday. Nice work!

**Pull request closed without merging**